src/server/stream_family.cc

```diff
@@ -2288,19 +2296,34 @@ void XReadImpl(CmdArgList args, std::optional<ReadOpts> opts, ConnectionContext*
         if (requested_sitem.id.val.ms != UINT64_MAX || requested_sitem.id.val.seq != UINT64_MAX) {
           block = false;
           opts->serve_history = true;
-          continue;
+          // continue;
+        } else {
+          requested_sitem.id.val = requested_sitem.group->last_id;
+          streamIncrID(&requested_sitem.id.val);
         }
-        requested_sitem.id.val = requested_sitem.group->last_id;
-        streamIncrID(&requested_sitem.id.val);
+        continue;
       }
 
       if (streamCompareID(&last_id, &requested_sitem.id.val) >= 0) {
         block = false;
       }
+    } else {
+      // if the key is not found,
+      // return NoGroupOrKey error.
+      // We are simply mimicking Redis' error message here...
+      // However, we could actually report more precise error message...
+      string error_msg = NoGroupOrKey(stream, opts->group_name);
+      if (opts->read_group) {
+        error_msg += " in XREADGROUP with GROUP option";
+      }
+      (*cntx)->SendError(error_msg);
+      return;
     }
   }
 
+  block = true;
   if (block) {
+    //(*cntx)->SendError("BLOCK SET!");
```

Comment:
We can remove this comment (or replace this with more helpful comment)

Reply:
fixed